### PR TITLE
[LF] drop retroactive interface instance

### PR DESCRIPTION
**Pull request by remyhaemmerle-da**

### CODE REVIEW

File: daml-lf/validation/src/test/scala/com/digitalasset/daml/lf/validation/TypingSpec.scala [hunk 1]
```diff
@@ -879,10 +878,7 @@ class TypingSpec(majorLanguageVersion: LanguageMajorVersion)
             val TTyCon(conI) = t"Mod:I"
             val TTyCon(conTi) = t"Mod:Ti"
             env.pkgInterface.lookupInterfaceInstance(conI, conTi) should matchPattern {
-              case Right(iiInfo: PackageInterface.InterfaceInstanceInfo)
-                  if iiInfo.interfaceId == conI
-                    && iiInfo.templateId == conTi
-                    && iiInfo.parent == TemplateOrInterface.Template(conTi) =>
+              case Right(TemplateImplementsSignature(`conI`, _)) =>
```
Comment (akrmn) [inline daml-lf/validation/src/test/scala/com/digitalasset/daml/lf/validation/TypingSpec.scala:881]: TIL about this use of backticks

File: daml-lf/validation/src/test/scala/com/digitalasset/daml/lf/validation/TypingSpec.scala [hunk 2]
```diff
@@ -966,27 +959,31 @@ class TypingSpec(majorLanguageVersion: LanguageMajorVersion)
...
-        E"""λ (t: Mod:Ti) → ⸨ to_interface @Mod:I @Mod:T t  ⸩""" -> //
-          { case EMissingInterfaceInstance(_, _, _) => },
+// Why ?
+//        E"""λ (t: Mod:Ti) → ⸨ to_interface @Mod:T @Mod:Ti t  ⸩""" -> //
+//          {
+//            case EUnknownDefinition(
+//                  _,
+//                  LookupError.NotFound(Reference.Interface(_), Reference.InterfaceInstance(_, _)),
+//                ) =>
+//          },
```
Comment (akrmn) [inline daml-lf/validation/src/test/scala/com/digitalasset/daml/lf/validation/TypingSpec.scala:969]: what's the question here? this test case used to fail because it's trying to use `to_interface` to convert into `Mod:T`, which isn't an interface. Since you changed the definition of `checkUniqueInterfaceInstanceExists`, the error might be different now

File: daml-lf/validation/src/test/scala/com/digitalasset/daml/lf/validation/TypingSpec.scala [hunk 3]
```diff
@@ -966,27 +959,31 @@ class TypingSpec(majorLanguageVersion: LanguageMajorVersion)
...
+// Why ?
+//        E"""λ (t: Mod:Ti) → ⸨ to_interface @Mod:I @Mod:T t  ⸩""" -> //
+//          { case EMissingInterfaceInstance(_, _, _) => },
```
Comment (akrmn) [inline daml-lf/validation/src/test/scala/com/digitalasset/daml/lf/validation/TypingSpec.scala:972]: this one expects the error because `Mod:T` does not implement `Mod:I` (which is confusing because it's called on a `Mod:Ti`). Why did you need to comment it out?

File: daml-lf/validation/src/test/scala/com/digitalasset/daml/lf/validation/TypingSpec.scala [hunk 4]
```diff
@@ -966,27 +959,31 @@ class TypingSpec(majorLanguageVersion: LanguageMajorVersion)
...
+//        E"""λ (i: Mod:I) → ⸨ from_interface @Mod:I @Mod:I i ⸩""" -> //
+//          {
+//            case EUnknownDefinition(
+//                  _,
+//                  LookupError.NotFound(Reference.Template(_), Reference.InterfaceInstance(_, _)),
+//                ) =>
+//          },
+// Why ?
+//        E"λ (i: Mod:I) → ⸨ from_interface @Mod:I @Mod:T i ⸩" -> //
+//          { case EMissingInterfaceInstance(_, _, _) => },
```
Comment (akrmn) [inline daml-lf/validation/src/test/scala/com/digitalasset/daml/lf/validation/TypingSpec.scala:986]: these two should behave as the previous two except they change `to_interface=>from_interface`

**Review by akrmn — APPROVED** [review]
Thanks @remyhaemmerle-da, looks good except for some commented-out test cases where I replied

**Review by carrielaben-da — COMMENTED** [review]
Looks good from a Docs perspective.

File: daml-lf/validation/src/test/scala/com/digitalasset/daml/lf/validation/TypingSpec.scala (hunk 2 above)
Comment (remyhaemmerle-da) [reply daml-lf/validation/src/test/scala/com/digitalasset/daml/lf/validation/TypingSpec.scala:969]: I just broke the type checker.  ;)
